libpas: Enable easier testing of PGM crashes

stwrt opened the pull request:
#### 0146f7f3441f8d788963818d223665fc231c74ff
<pre>
libpas: Enable easier testing of PGM crashes
<a href="https://bugs.webkit.org/show_bug.cgi?id=274863">https://bugs.webkit.org/show_bug.cgi?id=274863</a>

Reviewed by David Kilzer.

Add crash test scenarios to enable automated testing of PGM crashes in libpas for ReportCrash.
JavaScriptCore is required to perform testing for ReportCrash, so I added it to the jsc binary.

The new options available are:
./jsc --crash-vm=PGMOOBLowerGuardPage
./jsc --crash-vm=PGMOOBUpperGuardPage
./jsc --crash-vm=PGMUAF

* Source/JavaScriptCore/jsc.cpp:
(printUsageStatement):
(CommandLine::parseArguments):
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::forceEnablePGM):
* Source/WTF/wtf/FastMalloc.h:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_initialize_pgm_as_enabled):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:

Canonical link: <a href="https://commits.webkit.org/280443@main">https://commits.webkit.org/280443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3724b14d2ff4bc83108dbf1b10defd9e3d2e1775

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56591 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35918 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9064 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60205 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7032 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7223 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45843 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4922 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58621 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33791 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48842 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26704 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30551 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6174 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6035 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/49683 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52508 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6445 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61887 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55832 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/503 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6549 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53101 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/504 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48901 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53092 "Found 3 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/http-test-page-list (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12533 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/434 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/77593 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31746 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12857 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32832 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33916 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32578 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer reviewed (commented):
I really think we should figure out if we can put this code in `libWebCoreTestSupport.dylib` so we don't ship this test code.

We could also introduce `libJavaScriptCoreTestSupport.dylib` if that would be easier or simpler.

ddkilzer commented on Source/WTF/wtf/Assertions.cpp:
```diff
@@ -700,6 +704,35 @@ void WTFCrashWithInfoImpl(int, const char*, const char*, int, uint64_t) { CRASH(
+    *lowerGuardPage = 'a';
+
+    __builtin_unreachable();
+}
```
Would also be great to create both an OOB overflow and OOB underflow test functions since detection is different for each one.

stwrt replied:
Added support for under and over flow tests.

ddkilzer commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -4049,6 +4053,27 @@ void CommandLine::parseArguments(int argc, char** argv)
+            WTF::forceEnablePGM();
+            char* result = static_cast<char*>(fastMalloc(10000000));
+            result = result - getpagesize();
+            *result = 'a';
```
Let's extract this into a `static NEVER_INLINE` function so that it gets a crash signature like "`jsc:  test_pgm_oob_underflow`":

```
static NEVER_INLINE NO_RETURN_DUE_TO_CRASH void test_pgm_oob_underflow()
{
    WTF::forceEnablePGM();
    char* result = static_cast<char*>(fastMalloc(10000000));
    result = result - getpagesize();
    *result = 'a';
    RELEASE_ASSERT_NOT_REACHED();
}
```

stwrt replied:
Good point; let me extract these to new functions.

ddkilzer commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -4049,6 +4053,27 @@ void CommandLine::parseArguments(int argc, char** argv)
+            size_t allocSize = getpagesize() * 10000;
+            char* result = static_cast<char*>(fastMalloc(10000000));
+            result = result + allocSize;
+            *result = 'a';
```
Let's extract this into a `static NEVER_INLINE` function so that it gets a crash signature like "`jsc:  test_pgm_oob_overflow`":

```
static NEVER_INLINE NO_RETURN_DUE_TO_CRASH void test_pgm_oob_overflow()
{
    WTF::forceEnablePGM();
    size_t allocSize = getpagesize() * 10000;
    char* result = static_cast<char*>(fastMalloc(10000000));
    result = result + allocSize;
    *result = 'a';
    RELEASE_ASSERT_NOT_REACHED();
}
```

ddkilzer commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -4049,6 +4053,27 @@ void CommandLine::parseArguments(int argc, char** argv)
+            WTF::forceEnablePGM();
+            char* result = static_cast<char*>(fastMalloc(10000000));
+            fastFree(result);
+            *result = 'a';
```
Let's extract this into a `static NEVER_INLINE` function so that it gets a crash signature like "`jsc:  test_pgm_uaf`":

```
static NEVER_INLINE NO_RETURN_DUE_TO_CRASH void test_pgm_uaf()
{
    WTF::forceEnablePGM();
    char* result = static_cast<char*>(fastMalloc(10000000));
    fastFree(result);
    *result = 'a';
    RELEASE_ASSERT_NOT_REACHED();
}
```

ddkilzer reviewed (approved):
r=me if we extract `static` functions to force a human-readable crash signature.

BTW, why do these allocate 10,000,000 bytes?  Is that to force the "large allocation" code path?

stwrt committed:
libpas: Enable easier testing of PGM crashes
https://bugs.webkit.org/show_bug.cgi?id=274863

Reviewed by David Kilzer.

Add crash test scenarios to enable automated testing of PGM crashes in libpas for ReportCrash.
JavaScriptCore is required to perform testing for ReportCrash, so I added it to the jsc binary.

The new options available are:
./jsc --crash-vm=PGMOOBLowerGuardPage
./jsc --crash-vm=PGMOOBUpperGuardPage
./jsc --crash-vm=PGMUAF

* Source/JavaScriptCore/jsc.cpp:
(printUsageStatement):
(CommandLine::parseArguments):
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::forceEnablePGM):
* Source/WTF/wtf/FastMalloc.h:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_initialize_pgm_as_enabled):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:

Canonical link: https://commits.webkit.org/280443@main
